Genetics rework - Part 1

SirRichardFrancis opened the pull request:
## About The Pull Request

Replaces old genetics with what main design doc describes, albeit without domino minigame.
~This PR is work-in-progress, only 17~ 35 mutations is ready ~and new machinery is missing some features~.

![lab](https://user-images.githubusercontent.com/65828539/164965444-d3d965e7-2ab6-4f7c-9346-6401de24c936.png)


<details>
<summary>
Mutations
</summary>

Each human spawns with 3 random inactive mutations.
Cruciform slowly removes mutations from wearer, unless suppressed by "God Blood" mutation.
NeoTheology obelisks deal burn damage to mutants in sight, amount depends on quantity and tier of active mutations.
Currently available mutations:

### Tier 0 - Nero

Limit of 5 per person. Picked with 45% chance.

Blindness - Vision restricted to half tile away from the mob. 0 NSA.

Deafness - no sounds and most chat messages. 0 NSA.

Myopia - can't see too far without glasses. 0 NSA.

Stat debuffs - 6 mutations, one for each stat type. When active - decreases relevant stat by 15 point. 0 NSA.

### Tier 1 - Vespasian

Limit of 4 per person. Picked with 25% chance.

Stat buffs - 6 mutations, one for each stat type. When active - increases relevant stat by 25 point. 10 NSA.

No reject - Allows to transfuse any blood and attach any organs/limbs without consequences*. 10 NSA.


### Tier 2 - Tacitus

Limit of 3 per person. Picked with 15% chance.

Lesser healing - heal with effectiveness of tricordrazine, gives 15 NSA.

Mind projection - telepathically speak to other people. 15 NSA.

Remote observation - see trough other people's eyes. 15 NSA. 1 minute cooldown.

Force speak - force person in your line of sight to speak. 70% chance they'll know who did that. 15 NSA.

Numb fingers - leave no fingerprints. 0 NSA.

Roach pheromones - make roaches in your line of sight friendly for 1 minute. They will fight other roaches for you, but humans will lose sanity if they wear no internals or gasmask at the moment of ability activation. 15 NSA. 2 minutes cooldown.

Spider pheromones - same as roach pheromones, but for spiders. 15 NSA. 2 minutes cooldown.

### Tier 3 - Hadrian

Limit of 2 per person. Picked with 10% chance.

Reject - periodically rejects foreign bodies, be it a shrapnel, carrion spiders or an augmentation (crucifrom included).
If "No reject" is not present, doubles toxins from blood incompatibility. 20 NSA.

Greater healing - combined (not stacked) effects of kelotane, bicaridine (no percentage healing) and tricordrazine. 20 NSA.

Oborin Syndrome - comes with the perk, makes that condition curable/obtainable midgame. 0 NSA.

Night vision - effect of night vision googles. 20 NSA.

Thermal vision - effect of thermal lenses, no red filter. 20 NSA.

Inner fuhrer - imitate fuhrer roach's distress call. 20 NSA. 1 minute cooldown.

Telekinesis - classic SS13 telekinesis, but without the halo.

Clerical psi shielding - immunity to NT litanies (helpful ones too) and no damage from obelisks. 20 NSA.

### Tier 4 - Aurelien

Limit of 1 per person. Picked with 5% chance.

God Blood - Suppresses cruciform, allowing to have any implant or organ, as well as mutations.
Carrion with that mutation could be baptized with no negative consequences, other than every cruciformed on the ship being notified about that fact two minutes later.
Also suppresses excelsior compliant implant, allowing mutant to use Excelsior comms and be recognized by turrets/HUDs, while not actually being converted. 0 NSA.

X-ray vision - "Well secluded, I see all." 15 NSA.

Phazing - Ability to phase trough solid objects, be it a wall or something else. Takes 15 seconds. 20 NSA.

Morph body - Change eye/skin/hair/beard color, gender and haircut. Costs 50 units of blood. 10 NSA.

</details>
<details>
<summary>
Machinery
</summary>


### Gene regurgitator, aka moecubes printer

![printer_idle_yellow](https://user-images.githubusercontent.com/65828539/164965629-0df8faa6-6173-41a5-a486-2565d5592622.png)![printer_idle_cyan](https://user-images.githubusercontent.com/65828539/164965648-90f67c45-9e3d-4635-8c21-bbaef5ed60e4.png)![printer_idle_purple](https://user-images.githubusercontent.com/65828539/164965651-d57ab7d0-eef2-49f2-88bf-5e695bfa94b1.png)![printer_on_yellow](https://user-images.githubusercontent.com/65828539/164965631-ca832fba-3f7f-4671-b65b-ced14d961f5b.gif)![printer_on_cyan](https://user-images.githubusercontent.com/65828539/164965641-7d8c7dad-4931-418c-a820-538aca1e2344.gif)![printer_on_purple](https://user-images.githubusercontent.com/65828539/164965643-bd9ec363-bb5c-43c4-b157-1171360071a2.gif)![printer_error](https://user-images.githubusercontent.com/65828539/164965662-9a94961f-04f2-4f6c-98af-7d18f7fe4a47.gif)

![printer](https://user-images.githubusercontent.com/65828539/164965326-8f8776b8-cad6-41b3-b12f-dd98937d2629.png)


That thing accepts data discs and a beaker, and prints out three types of disgusting fleshy cubes, which is Eris analog of dna injectors. Cube costs 15 units of nutriments (nutriment, protein, glucose or egg) regardless of type.
Does not accept inactive/dormant mutations.

Cube types:

1) Meat cube. Transfers whatever mutation/data is loaded in it to whoever eat it. Made by pressing "Make imprinter".![genecube](https://user-images.githubusercontent.com/65828539/164965900-d9cc0e55-34e9-462a-9137-2f5540e003a8.gif)

2) Worm cube with a mutation. Removes a very specific mutation from the target. Made by pressing "Make antidot" with a mutation selected.![wormcube](https://user-images.githubusercontent.com/65828539/164966025-bffca5a0-d92d-4be3-8291-c249dcead7e2.gif)

3) Worm cube without a mutation. Removes random mutation from the target. Made by pressing the same button, but when nothing/non-mutation is selected.![wormcube](https://user-images.githubusercontent.com/65828539/164966025-bffca5a0-d92d-4be3-8291-c249dcead7e2.gif)


### Chrysalis pod, aka stationary dna scanner

![cybercoffin_closed](https://user-images.githubusercontent.com/65828539/164966418-ebe44818-0d5a-4f7f-86c6-3bc8df33e69b.png)![cybercoffin_open](https://user-images.githubusercontent.com/65828539/164966421-5f7f8d81-8cf9-4665-8f33-6683ad918c4b.png)

Put some guy in there and use console nearby.

### Chrysalis controller, aka dna scanner console
![console_idle_yellow](https://user-images.githubusercontent.com/65828539/164966327-6806de74-8b6c-46e6-9716-84759ff67676.gif)![console_idle_cyan](https://user-images.githubusercontent.com/65828539/164966330-b3b99c55-f850-46a5-926b-58a015d19c4d.gif)![console_idle_purple](https://user-images.githubusercontent.com/65828539/164966339-86258617-cb3f-410e-b708-198bbba0df99.gif)![console_on_yellow](https://user-images.githubusercontent.com/65828539/164966343-b570ba37-cda9-4cbd-b819-0b93de8b3d97.gif)![console_on_cyan](https://user-images.githubusercontent.com/65828539/164966346-abce312b-56c4-4e88-b348-135d56cf09a6.gif)![console_on_purple](https://user-images.githubusercontent.com/65828539/164966347-7f4b553b-ddb3-4e91-8fd0-d8adac872dee.gif)![console_error](https://user-images.githubusercontent.com/65828539/164966351-bbf9b2fd-015f-412d-8dcc-b57ad8e6edfb.gif)

![console](https://user-images.githubusercontent.com/65828539/164964665-a1d10406-9dbf-4800-a897-569e39160944.png)

Transfers mutations or mob's unique appearance, dna trace, fingerprints and a blood type to a disc.

### ML "Sensillum", aka portable dna scanner

![dnascanner_working_yellow](https://user-images.githubusercontent.com/65828539/164967043-c327586d-d60e-4553-97a2-417ede6c4cf0.gif)![dnascanner_no_yellow](https://user-images.githubusercontent.com/65828539/164967048-c8f15fc0-7b03-49ad-a27d-24faf47deead.gif)![dnascanner_working_cyan](https://user-images.githubusercontent.com/65828539/164967059-60737de7-3e6b-4118-8fe0-e6cb7a755d3a.gif)![dnascanner_no_cyan](https://user-images.githubusercontent.com/65828539/164967062-d2d5e67f-4cb0-47fc-b8f9-70d01fe7d1ca.gif)![dnascanner_working_purple](https://user-images.githubusercontent.com/65828539/164967078-ea47196d-98b5-4e36-8ee0-53cc6da62d6b.gif)![dnascanner_no_purple](https://user-images.githubusercontent.com/65828539/164967079-74891fb5-bc43-4625-b45f-d2743d8b9412.gif)

Serves the same purpose as stationary scanner, but requires a new chemical - Kognim to be able to scan mutations.
Kognim is made out of one part radium, two parts ryetalyn and one part synaptizine.
Causes constant pain to the patient, so injection of more than 1 unit is unadvised.
Injection of 5 units with no painkillers and basic stats will cause patient to go into paincrit after a minute.

### Genome decoder

A program for modular laptops and consoles. Pick an inactive mutation from disc (portable or not) and wait for it to be turned into an active one. Time depends on computer's processing power and amount of programs running in the background.
That's it. A placeholder. Will be replaced with better system in following PR.


On a side note, most of new machinery have several color schemes that is randomly picked at roundstart.
Color could be changed with use of multitool.

https://user-images.githubusercontent.com/65828539/162403159-4bfb4113-6de2-49f1-89e7-30f8903ae9e8.mp4

Those are also a part of the tech tree.

![image](https://user-images.githubusercontent.com/65828539/164967636-ec66f6dd-f2cc-476e-96b0-4281977c7856.png)


</details>

## Why It's Good For The Game

You know why.

## Changelog
:cl:
refactor: dna, mutations and genetics.
/:cl:

SirRichardFrancis commented:
https://user-images.githubusercontent.com/65828539/162403159-4bfb4113-6de2-49f1-89e7-30f8903ae9e8.mp4

TheShown911 commented:
Few runs on the test server would be good.

SirRichardFrancis commented:
In case someone wondering why integration test failed:
![image](https://user-images.githubusercontent.com/65828539/162406208-92abf7ea-f8b6-452b-92d4-9adc43de2b42.png)

MLGTASTICa commented on code/datums/perks/fate.dm:
```diff
@@ -214,12 +214,19 @@
+			M.imprint(holder)
 		spawn(1)
-			holder.update_client_colour() //Handle the activation of the colourblindness on the mob.
+			holder?.update_client_colour() //Handle the activation of the colourblindness on the mob.
```
why ?. , this should 100% runtime if the holder is actually null. because imaking applying perks to null..

SirRichardFrancis replied:
Perk is not applied to null holder, but holder may be gone when `update_client_colour()` is called.
It does not runtime with `?.`, but does without.

MLGTASTICa replied:
why would the holder be gone ?

SirRichardFrancis replied:
When perk is added or removed, it calls for related mutation to be added/removed as well, and vice versa.
That causes perk addition/removal to being called twice (initial and from mutation).
Addition of `?.` is simplest way to prevent runtime. Could be replaced with extra checks, but dunno if that's necessary.

SirRichardFrancis replied:
Actually null holder is not the case when perk is added, only when removed.

MLGTASTICa commented on code/modules/genetics/moecubes_printer.dm:
```diff
@@ -0,0 +1,182 @@
+
+
+/obj/machinery/moeballs_printer/proc/try_eject_usb(mob/user)
+	if(usb && eject_item(usb, user))
```
does this check anywhere is the user is a ghost or is not alive (stat)

SirRichardFrancis replied:
Added checks for `Adjacent()` and `user.stat`. Ghosts fail `IsAdvancedToolUser()`.

MLGTASTICa commented on code/modules/genetics/moecubes_printer.dm:
```diff
@@ -0,0 +1,182 @@
+
+	if(href_list["make_worm_cube"])
+		do_flick(FALSE)
+		sleep(1.5 SECONDS)
```
your using both sleep and then returning topic_Refresh , whats the point ? the update will be delayed either way .

SirRichardFrancis replied:
It takes 1.5 seconds for cube printing animation to play, hence the spawning delay.

MLGTASTICa replied:
fair

MLGTASTICa commented on code/modules/genetics/stationary_scanner.dm:
```diff
@@ -0,0 +1,124 @@
+
+
+/obj/machinery/cryo_slab/MouseDrop_T(mob/target, mob/user)
+	if(!ismob(target))
```
did you check if ghosts canțt just drag anyone in it ?

SirRichardFrancis replied:
They can't.

MLGTASTICa commented on code/modules/genetics/stationary_scanner_console.dm:
```diff
@@ -0,0 +1,303 @@
+/obj/machinery/dna_console/New()
+	..()
+	color_key = default_dna_machinery_style
+	establish_connection()
```
consider adding a verb for manual connections / disconnections , so people can rebuild or remove it

MLGTASTICa commented on code/modules/genetics/stationary_scanner_console.dm:
```diff
@@ -0,0 +1,303 @@
+		..()
+
+
+/obj/machinery/dna_console/AltClick(mob/user)
```
ghost validation ? or if the user is dead / incapacitated?

MLGTASTICa commented on code/modules/mining/abandonedcrates.dm:
```diff
@@ -94,7 +94,7 @@
 		if(91)
 			new/obj/item/tool/sword/katana(src)
 		if(92)
-			new/obj/item/dnainjector/xraymut(src) // Probably the least OP
```
least based change,

SirRichardFrancis replied:
Dna injectors had to go.

MLGTASTICa reviewed (commented):
looks fine in general , except the removal of the xray mut from the abandoned crates , cringe , unbased and tgpilled :(

SirRichardFrancis commented on code/modules/genetics/portable_scanner.dm:
```diff
@@ -0,0 +1,109 @@
+
+
+/obj/item/dna_scanner/proc/try_eject_usb(mob/user)
+	if(usb && eject_item(usb, user))
```
Addressed.

SirRichardFrancis commented:
> Doesn't look like the God Blood trait suppresses things like baptism effects (big one being nonhuman/carrion baptism gib), any chance you could change that?

Indeed.

SirRichardFrancis commented:
Updated tech tree.
![image](https://user-images.githubusercontent.com/65828539/162712496-a35a4906-8a3f-43da-adb8-ae1ef70f2bc7.png)

SirRichardFrancis commented:
> Are these obtainable via junk?

Yes, although chances of obtaining the right items is low.

SirRichardFrancis commented:
Added/fixed cruciform and excelsior implant interactions and implemented these suggestions:
![image](https://user-images.githubusercontent.com/65828539/163038994-7a9e2302-bc74-497f-ba64-70fd915a7c03.png)
